File: go-wallet/domain_whitelist.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// DomainWhitelistEntry represents a whitelisted domain
type DomainWhitelistEntry struct {
	Domain       string    `json:"domain"`
	AddedAt      time.Time `json:"addedAt"`
	LastUsed     time.Time `json:"lastUsed"`
	RequestCount int       `json:"requestCount"`
	IsPermanent  bool      `json:"isPermanent"` // true for whitelist, false for one-time
}

// DomainWhitelistManager manages the domain whitelist
type DomainWhitelistManager struct {
	whitelist map[string]*DomainWhitelistEntry
	mutex     sync.RWMutex
	filePath  string
}

// NewDomainWhitelistManager creates a new domain whitelist manager
func NewDomainWhitelistManager() *DomainWhitelistManager {
	homeDir, _ := os.UserHomeDir()
	filePath := filepath.Join(homeDir, "AppData", "Roaming", "BabbageBrowser", "wallet", "domainWhitelist.json")

	manager := &DomainWhitelistManager{
		whitelist: make(map[string]*DomainWhitelistEntry),
		filePath:  filePath,
	}

	// Load existing whitelist
	manager.loadWhitelist()

	return manager
}

// loadWhitelist loads the whitelist from file
func (dwm *DomainWhitelistManager) loadWhitelist() error {
	dwm.mutex.Lock()
	defer dwm.mutex.Unlock()

	// Create directory if it doesn't exist
	dir := filepath.Dir(dwm.filePath)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return fmt.Errorf("failed to create directory: %v", err)
	}

	// Check if file exists
	if _, err := os.Stat(dwm.filePath); os.IsNotExist(err) {
		// File doesn't exist, start with empty whitelist
		return nil
	}

	// Read file
	data, err := os.ReadFile(dwm.filePath)
	if err != nil {
		return fmt.Errorf("failed to read whitelist file: %v", err)
	}

	// Parse JSON
	var entries []DomainWhitelistEntry
	if err := json.Unmarshal(data, &entries); err != nil {
		return fmt.Errorf("failed to parse whitelist file: %v", err)
	}

	// Convert to map
	for _, entry := range entries {
		dwm.whitelist[entry.Domain] = &entry
	}

	return nil
}

// saveWhitelist saves the whitelist to file
// Note: Caller must hold the mutex
func (dwm *DomainWhitelistManager) saveWhitelist() error {

	// Convert map to slice
	var entries []DomainWhitelistEntry
	for _, entry := range dwm.whitelist {
		entries = append(entries, *entry)
	}

	// Marshal to JSON
	data, err := json.MarshalIndent(entries, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to marshal whitelist: %v", err)
	}

	// Write to file
	if err := os.WriteFile(dwm.filePath, data, 0644); err != nil {
		return fmt.Errorf("failed to write whitelist file: %v", err)
	}

	return nil
}

// IsDomainWhitelisted checks if a domain is whitelisted
func (dwm *DomainWhitelistManager) IsDomainWhitelisted(domain string) bool {
	dwm.mutex.RLock()
	defer dwm.mutex.RUnlock()

	entry, exists := dwm.whitelist[domain]
	if !exists {
		return false
	}

	// Check if it's a one-time entry that has been used
	if !entry.IsPermanent && entry.RequestCount > 0 {
		return false
	}

	return true
}

// AddToWhitelist adds a domain to the whitelist
func (dwm *DomainWhitelistManager) AddToWhitelist(domain string, isPermanent bool) error {
	dwm.mutex.Lock()
	defer dwm.mutex.Unlock()

	now := time.Now()
	entry := &DomainWhitelistEntry{
		Domain:       domain,
		AddedAt:      now,
		LastUsed:     now,
		RequestCount: 0,
		IsPermanent:  isPermanent,
	}

	dwm.whitelist[domain] = entry

	// Save to file
	return dwm.saveWhitelist()
}

// RecordRequest records a request from a domain
func (dwm *DomainWhitelistManager) RecordRequest(domain string) error {
	dwm.mutex.Lock()
	defer dwm.mutex.Unlock()

	entry, exists := dwm.whitelist[domain]
	if !exists {
		return fmt.Errorf("domain not in whitelist: %s", domain)
	}

	entry.LastUsed = time.Now()
	entry.RequestCount++

	// Save to file
	return dwm.saveWhitelist()
}

// RemoveFromWhitelist removes a domain from the whitelist
func (dwm *DomainWhitelistManager) RemoveFromWhitelist(domain string) error {
	dwm.mutex.Lock()
	defer dwm.mutex.Unlock()

	delete(dwm.whitelist, domain)

	// Save to file
	return dwm.saveWhitelist()
}

// GetWhitelist returns the current whitelist
func (dwm *DomainWhitelistManager) GetWhitelist() map[string]*DomainWhitelistEntry {
	dwm.mutex.RLock()
	defer dwm.mutex.RUnlock()

	// Return a copy
	result := make(map[string]*DomainWhitelistEntry)
	for domain, entry := range dwm.whitelist {
		result[domain] = entry
	}

	return result
}

// GetWhitelistAsJSON returns the whitelist as JSON for API responses
func (dwm *DomainWhitelistManager) GetWhitelistAsJSON() ([]byte, error) {
	dwm.mutex.RLock()
	defer dwm.mutex.RUnlock()

	var entries []DomainWhitelistEntry
	for _, entry := range dwm.whitelist {
		entries = append(entries, *entry)
	}

	return json.MarshalIndent(entries, "", "  ")
}

File: rust-wallet/src/domain_whitelist.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;
use std::sync::{Arc, Mutex};
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DomainWhitelistEntry {
    pub domain: String,
    pub added_at: i64,      // Unix timestamp (seconds since epoch)
    pub last_used: i64,     // Unix timestamp (seconds since epoch)
    pub request_count: u32,
    pub is_permanent: bool,
}

pub struct DomainWhitelistManager {
    whitelist: Arc<Mutex<HashMap<String, DomainWhitelistEntry>>>,
    file_path: PathBuf,
}

impl DomainWhitelistManager {
    pub fn new() -> Self {
        let file_path = Self::get_whitelist_path();

        let manager = Self {
            whitelist: Arc::new(Mutex::new(HashMap::new())),
            file_path,
        };

        // Load existing whitelist
        if let Err(e) = manager.load_whitelist() {
            log::warn!("Failed to load whitelist: {}", e);
        }

        manager
    }

    fn get_whitelist_path() -> PathBuf {
        let home_dir = dirs::home_dir().expect("Could not determine home directory");
        home_dir
            .join("AppData")
            .join("Roaming")
            .join("HodosBrowser")
            .join("wallet")
            .join("domainWhitelist.json")
    }

    fn load_whitelist(&self) -> Result<(), Box<dyn std::error::Error>> {
        // Create directory if it doesn't exist
        if let Some(parent) = self.file_path.parent() {
            fs::create_dir_all(parent)?;
        }

        // Check if file exists
        if !self.file_path.exists() {
            log::info!("📂 Whitelist file does not exist yet: {:?}", self.file_path);
            return Ok(());
        }

        // Read file
        let data = fs::read_to_string(&self.file_path)?;
        log::info!("📂 Loading whitelist from: {:?}", self.file_path);

        // Parse JSON
        let entries: Vec<DomainWhitelistEntry> = serde_json::from_str(&data)?;
        log::info!("📂 Loaded {} existing domains from whitelist", entries.len());

        // Convert to map
        let mut whitelist = self.whitelist.lock().unwrap();
        for entry in entries {
            log::info!("   - {}", entry.domain);
            whitelist.insert(entry.domain.clone(), entry);
        }

        Ok(())
    }

    fn save_whitelist(&self) -> Result<(), Box<dyn std::error::Error>> {
        let whitelist = self.whitelist.lock().unwrap();

        // Convert map to vec
        let entries: Vec<DomainWhitelistEntry> = whitelist.values().cloned().collect();
        log::info!("💾 Saving {} domains to whitelist file: {:?}", entries.len(), self.file_path);
        for entry in &entries {
            log::info!("   - {}", entry.domain);
        }

        // Marshal to JSON
        let data = serde_json::to_string_pretty(&entries)?;

        // Write to file
        fs::write(&self.file_path, data)?;
        log::info!("✅ Whitelist saved successfully");

        Ok(())
    }

    pub fn is_domain_whitelisted(&self, domain: &str) -> bool {
        let whitelist = self.whitelist.lock().unwrap();

        if let Some(entry) = whitelist.get(domain) {
            // Check if it's a one-time entry that has been used
            if !entry.is_permanent && entry.request_count > 0 {
                return false;
            }
            return true;
        }

        false
    }

    pub fn add_to_whitelist(&self, domain: String, is_permanent: bool) -> Result<(), Box<dyn std::error::Error>> {
        let mut whitelist = self.whitelist.lock().unwrap();

        log::info!("📝 Adding domain to whitelist: {} (permanent: {})", domain, is_permanent);
        log::info!("📝 Current whitelist size before add: {}", whitelist.len());

        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs() as i64;

        let entry = DomainWhitelistEntry {
            domain: domain.clone(),
            added_at: now,
            last_used: now,
            request_count: 0,
            is_permanent,
        };

        whitelist.insert(domain.clone(), entry);
        log::info!("📝 Current whitelist size after add: {}", whitelist.len());
        drop(whitelist); // Release lock before saving

        self.save_whitelist()?;

        log::info!("✅ Added domain to whitelist: {} (permanent: {})", domain, is_permanent);

        Ok(())
    }

    pub fn record_request(&self, domain: &str) -> Result<(), Box<dyn std::error::Error>> {
        let mut whitelist = self.whitelist.lock().unwrap();

        if let Some(entry) = whitelist.get_mut(domain) {
            entry.last_used = SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap()
                .as_secs() as i64;
            entry.request_count += 1;
            drop(whitelist); // Release lock before saving

            self.save_whitelist()?;
            Ok(())
        } else {
            Err(format!("domain not in whitelist: {}", domain).into())
        }
    }

    pub fn remove_from_whitelist(&self, domain: &str) -> Result<(), Box<dyn std::error::Error>> {
        let mut whitelist = self.whitelist.lock().unwrap();
        whitelist.remove(domain);
        drop(whitelist); // Release lock before saving

        self.save_whitelist()?;

        log::info!("✅ Removed domain from whitelist: {}", domain);

        Ok(())
    }

    pub fn get_whitelist(&self) -> HashMap<String, DomainWhitelistEntry> {
        let whitelist = self.whitelist.lock().unwrap();
        whitelist.clone()
    }
}
